File: cache_interceptor/src/test/java/uk/co/glass_software/android/cache_interceptor/interceptors/error/ErrorInterceptorIntegrationTest.java
package uk.co.glass_software.android.cache_interceptor.interceptors.error;


import org.junit.Before;
import org.junit.Test;

import java.io.IOException;

import io.reactivex.Observable;
import uk.co.glass_software.android.cache_interceptor.base.BaseIntegrationTest;
import uk.co.glass_software.android.cache_interceptor.base.network.model.TestResponse;
import uk.co.glass_software.android.cache_interceptor.interceptors.cache.CacheToken;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertNotNull;
import static org.mockito.Mockito.mock;
import static uk.co.glass_software.android.cache_interceptor.interceptors.cache.CacheStatus.DO_NOT_CACHE;
import static uk.co.glass_software.android.cache_interceptor.interceptors.error.ApiErrorFactoryUnitTest.assertApiError;
import static uk.co.glass_software.android.cache_interceptor.interceptors.error.ErrorCode.NETWORK;

@SuppressWarnings("unchecked")
public class ErrorInterceptorIntegrationTest extends BaseIntegrationTest {
    
    private final String message = "Connection timed out";
    
    private ErrorInterceptor<TestResponse, ApiError> target;
    
    @Before
    public void setUp() throws Exception {
        CacheToken<TestResponse> cacheToken = CacheToken.Companion.newRequest(
                TestResponse.class,
                TestResponse.URL,
                null,
                5f
        );
        
        target = new ErrorInterceptor.Factory<>(
                new ApiErrorFactory(),
                mock(Logger.class)
        ).create(cacheToken);
    }
    
    @Test
    public void testInterceptorWithError() throws Exception {
        testInterceptor(true);
    }
    
    @Test
    public void testInterceptorNoError() throws Exception {
        testInterceptor(false);
    }
    
    private void testInterceptor(boolean hasError) throws Exception {
        TestResponse successResponse = new TestResponse();
        Observable<TestResponse> observable = Observable.just(successResponse);
        
        if (hasError) {
            observable = observable.doOnSubscribe(ignore -> {
                throw new IOException(message);
            });
        }
        
        TestResponse response = observable.compose(target).blockingFirst();
        
        ResponseMetadata<TestResponse, ApiError> metadata = response.getMetadata();
        assertNotNull("TestResponse should have metadata", metadata);
        
        if (hasError) {
            assertApiError(
                    metadata.getError(),
                    "Connection timed out",
                    NETWORK,
                    -1,
                    true
            );
            
            assertEquals("Cache token should be DoNotCache",
                         DO_NOT_CACHE,
                         metadata.getCacheToken().getStatus()
            );
        }
        else {
            assertFalse("PotListResponse should not have errors", metadata.hasError());
            assertEquals("Responses didn't match", successResponse, response);
        }
    }
    
}

File: cache_interceptor/src/test/java/uk/co/glass_software/android/dejavu/interceptors/internal/error/ErrorInterceptorIntegrationTest.kt
package uk.co.glass_software.android.dejavu.interceptors.internal.error

import io.reactivex.Observable
import org.junit.Before
import org.junit.Test
import uk.co.glass_software.android.dejavu.configuration.CacheInstruction.Operation.Expiring.Cache
import uk.co.glass_software.android.dejavu.injection.integration.component.IntegrationCacheComponent
import uk.co.glass_software.android.dejavu.injection.module.CacheModule
import uk.co.glass_software.android.dejavu.interceptors.internal.cache.token.CacheToken
import uk.co.glass_software.android.dejavu.test.*
import uk.co.glass_software.android.dejavu.test.network.model.TestResponse

internal class ErrorInterceptorIntegrationTest
    : BaseIntegrationTest<CacheModule.Function2<CacheToken, Long, ErrorInterceptor<Glitch>>>(IntegrationCacheComponent::errorInterceptorFactory) {

    private lateinit var targetErrorInterceptor: ErrorInterceptor<Glitch>

    @Before
    fun setUp() {
        targetErrorInterceptor = target.get(
                instructionToken(Cache()),
                1234L
        )
    }

    private fun apply(observable: Observable<Any>) =
            targetErrorInterceptor.apply(observable).blockingFirst()

    @Test
    fun testApplyWithResponse() {
        val value = Object()

        val result = apply(Observable.just(value))

        assertEqualsWithContext(
                TestResponse::class.java,
                result.responseClass,
                "Response class didn't match"
        )

        assertEqualsWithContext(
                TestResponse::class.java,
                result.responseClass,
                "Response didn't match"
        )

        val metadata = result.metadata

        assertNullWithContext(
                metadata.exception,
                "Metadata should not have an exception"
        )

        assertEqualsWithContext(
                instructionToken(Cache()),
                metadata.cacheToken,
                "Cache token didn't match"
        )
    }

    @Test
    fun testApplyWithEmpty() {
        testApplyWithException(
                Observable.empty(),
                NoSuchElementException("Response was empty")
        )
    }

    @Test
    fun testApplyWithError() {
        val error = NullPointerException("error")
        testApplyWithException(
                Observable.error(error),
                error
        )
    }

    private fun testApplyWithException(observable: Observable<Any>,
                                       exception: Throwable) {
        val result = apply(observable)

        assertEqualsWithContext(
                TestResponse::class.java,
                result.responseClass,
                "Response class didn't match"
        )

        assertNullWithContext(
                result.response,
                "Response should be null"
        )

        val metadata = result.metadata

        assertNotNullWithContext(
                metadata.exception,
                "Metadata should have an exception"
        )

        assertEqualsWithContext(
                true,
                metadata.exception is Glitch,
                "Exception should be Glitch"
        )

        val glitch = metadata.exception as Glitch

        assertGlitchWithContext(
                Glitch(exception),
                glitch,
                "Glitch didn't match"
        )

        assertEqualsWithContext(
                instructionToken(Cache()),
                metadata.cacheToken,
                "Cache token didn't match"
        )
    }
}